posix/select.c

```diff
@@ -0,0 +1,236 @@
+    }
+
+    {
+        const size_t num_bytes = sizeof(uint32_t) * (size_t)(nfds + 1);
+
+        if (oe_memset_s(fd_list, num_bytes, 0xff, num_bytes) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+    if (readfds)
+    {
+        _set_to_fd_list(
+            readfds,
+            (OE_EPOLLIN | OE_EPOLLRDNORM | OE_EPOLLRDBAND),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    if (writefds)
+    {
+        _set_to_fd_list(
+            writefds,
+            (OE_EPOLLOUT | OE_EPOLLWRNORM | OE_EPOLLWRBAND),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    if (exceptfds)
+    {
+        _set_to_fd_list(
+            exceptfds,
+            (OE_EPOLLERR | OE_EPOLLHUP | OE_EPOLLRDHUP | OE_EPOLLWAKEUP),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    epfd = oe_epoll_create1(0);
+    if (epfd < 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    for (i = 0; i < nfds; i++)
+    {
+        struct oe_epoll_event ev = {
+            .data.fd = fd_list[i],
+            .events = (uint32_t)fd_flags[fd_list[i]],
+        };
+
+        ret = oe_epoll_ctl(epfd, OE_EPOLL_CTL_ADD, fd_list[i], &ev);
+        if (ret < 0)
+        {
+            OE_RAISE_ERRNO_MSG(
+                OE_EINVAL, "epfd=%d fd_list[%d]=%d", epfd, i, fd_list[i]);
+        }
+    }
+
+    ret_fds = oe_epoll_wait(epfd, rtn_ev, OE_FD_SETSIZE, timeout_ms);
+    if (ret_fds < 0)
+    {
+        goto done;
+    }
+
+    if (readfds)
+        OE_FD_ZERO(readfds);
+
+    if (writefds)
+        OE_FD_ZERO(writefds);
+
+    if (exceptfds)
+        OE_FD_ZERO(exceptfds);
+
+    if (readfds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLIN | OE_EPOLLRDNORM | OE_EPOLLRDBAND),
+            readfds);
+    }
+
+    if (writefds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLOUT | OE_EPOLLWRNORM | OE_EPOLLWRBAND),
+            writefds);
+    }
+
+    if (exceptfds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLERR | OE_EPOLLHUP | OE_EPOLLRDHUP | OE_EPOLLWAKEUP),
+            exceptfds);
+    }
+done:
+
+    if (epfd >= 0)
+        oe_close(epfd);
+
+    return ret_fds;
+}
+
+void OE_FD_CLR(int fd, oe_fd_set* set)
+{
+    int l = fd >> 5;   // long index
+    int b = fd & 0x1f; // bit shift
+    set->fds_bits[l] &= ~(1UL << b);
+}
+
+int OE_FD_ISSET(int fd, oe_fd_set* set)
+{
+    int l = fd >> 5;   // long index
+    int b = fd & 0x1f; // bit shift
+    return (set->fds_bits[l] & (1UL << b)) != 0;
+}
+
+void OE_FD_SET(int fd, oe_fd_set* set)
+{
+    int l = fd >> 5;   // long index
+    int b = fd & 0x1f; // bit shift
+
+    set->fds_bits[l] |= (1UL << b);
+}
+
+void OE_FD_ZERO(oe_fd_set* set)
+{
+    int i = 0;
+    uint64_t* bits = set->fds_bits;
+
+    while ((uint32_t)i < (sizeof(oe_fd_set) / sizeof(uint64_t)))
```

Comment:
Nit: just declare `i` as `uint32_t` instead of casting.

Reply:
I simplified to this:

```
void OE_FD_ZERO(oe_fd_set* set)
{
    if (set)
    {
        /* oe_memset_s() cannot fail with these parameters. */
        oe_memset_s(set, sizeof(oe_fd_set), 0, sizeof(oe_fd_set));
    }
}
```